share_app/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -178,42 +192,77 @@ private void readForm() {
             int numOfRes = dis.readInt();
+            String formMediaPath = FORMS_PATH + "/" + displayName + "-media";
             while (numOfRes-- > 0) {
-                receiveFile(FORM_PATH + displayName + "-media");
+                receiveFile(formMediaPath);
             }
+
+            // Add row in forms db
+            ContentValues values = new ContentValues();
+            values.put(FormsProviderAPI.FormsColumns.FORM_FILE_PATH, FORMS_PATH + "/" + formName);
+            values.put(FormsProviderAPI.FormsColumns.DISPLAY_NAME, displayName);
+            values.put(FormsProviderAPI.FormsColumns.JR_FORM_ID, formId);
+            values.put(FormsProviderAPI.FormsColumns.JR_VERSION, formVersion);
+            values.put(FormsProviderAPI.FormsColumns.SUBMISSION_URI, submissionUri);
+            values.put(FormsProviderAPI.FormsColumns.FORM_MEDIA_PATH, formMediaPath);
+            new FormsDao().saveForm(values);
         } catch (IOException e) {
             Timber.e(e);
         }
     }
 
-    private void readInstances(String formId) {
+    private void readInstances(String formId, String formVersion) {
         try {
             int numInstances = dis.readInt();
             while (numInstances-- > 0) {
                 // publish current progress
                 rxEventBus.post(new DownloadEvent(DownloadEvent.Status.DOWNLOADING, ++progress, total));
+                String displayName = dis.readUTF();
+                String submissionUri = dis.readUTF();
+
+                if (submissionUri.equals("-1")) {
+                    submissionUri = null;
+                }
 
                 int numRes = dis.readInt();
                 String time = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss-SSS",
                         Locale.ENGLISH).format(Calendar.getInstance().getTime());
-                String path = INSTANCE_PATH + formId + "_" + time;
+                String path = INSTANCES_PATH + "/" + formId + "_" + time;
                 while (numRes-- > 0) {
                     receiveFile(path);
                 }
+
+                // Add row in instances table
+                ContentValues values = new ContentValues();
+                values.put(DISPLAY_NAME, displayName);
+                values.put(STATUS, InstanceProviderAPI.STATUS_COMPLETE);
+                values.put(CAN_EDIT_WHEN_COMPLETE, "true");
+                values.put(SUBMISSION_URI, submissionUri);
+                values.put(INSTANCE_FILE_PATH, path);
+                values.put(JR_FORM_ID, formId);
+                values.put(JR_VERSION, formVersion);
+                Uri uri = new InstancesDao().saveInstance(values);
+
+                // Add row in share table
+                ContentValues shareValues = new ContentValues();
+                shareValues.put(INSTANCE_ID, Long.parseLong(uri.getLastPathSegment()));
+                shareValues.put(TRANSFER_STATUS, STATUS_FORM_RECEIVE);
+                new ShareDatabaseHelper().insertInstance(shareValues);
```

Comment:
The above two values can be sent as a parameter to the `insertInstance` method

Reply:
I wrote like this because sometimes I may get a feedback from the sender and sometimes not. So I don't know when I'm gonna receive it, maybe in future, some more fields may get added to a database that's why I didn't send all values as a parameter. 
What do you think?

Reply:
We will be updating the table after review as well as after every instance transfer.  There are two cases one when instance is received other one is any user marks instance as reviewed after giving the feedback. But sometime user doesn't give any feedback so in that case we have to write multiple methods for each case.